src/nupic/algorithms/SpatialPooler.cpp

```diff
@@ -257,6 +257,7 @@ void SpatialPooler::setSynPermConnected(Real synPermConnected) {
   NTA_CHECK( synPermConnected > connections::minPermanence );
   NTA_CHECK( synPermConnected <= connections::maxPermanence );
   synPermConnected_ = synPermConnected;
+  connections_.setConnectedThreshold( synPermConnected_ );
```

Comment:
has the behavoir so far been broken without this functionality?

Reply:
The method `sp::setSynPermCon` is broken without this fix.  No one calls this method though.  This has been broken since the connections class optimization to presyn tables.

Reply:
> No one calls this method though.  

Question if we shouldn't instead remove the method; it does still get set by the init param, right?

Reply:
> it does still get set by the init param, right?

Correct.  The question is: is there any reason why the user might want to change the threshold while the program is running?  I don't think so.  There are many ways to manipulate the permanence values instead of moving the threshold, all of which can accomplish the same end result and which all have clear use cases.  Alternatives include: `bumpSegment`, `raisePermanencesToThreshold`, changing permanence increments/decrements, and we can add more methods when they're needed by a specific application/use-case.

The test cases which exercise this method which I've seen all create the connections using default arguments, and then immediately change this before running anything.  So the tests could be made to work without this method.

But! this is part of the public API so before removing we should do a search through all extant code for calls to `SP.setSynPermConnected` and `TM.setConnectedPermanence`.

Reply:
> Correct. The question is: is there any reason why the user might want to change the threshold while the program is running? I don't think so. There are many [other] ways to manipulate the permanence values 

I totally agree. Changing it at runtime would break all the learned values, and even then there are ways to achieve the behavior.. 
Therefore, I'd vote to remove the method, instead of adding a helper.

Reply:
@dkeeney, do you have an opinion on removing these methods from public API?

Reply:
> @dkeeney, do you have an opinion on removing these methods from public API?

Anything we remove could cause us compatibility problems later but if you don't see anyone calling this from Python it is probably safe to document the change and take it out.

Reply:
Can we somehow proceed with this PR? 
- either implement this PR and fix the method
- or I'd prefer to remove the broken method
> The method sp::setSynPermCon is broken without this fix. No one calls this method though. This has been broken since the connections class optimization to presyn tables. 

And remove from public API the callers
> But! this is part of the public API so before removing we should do a search through all extant code for calls to SP.setSynPermConnected and TM.setConnectedPermanence.

This will not be a heavily used feature, and if used, it can be easily replaced.

Reply:
Yes we should remove this method.  I'd like for this PR to also fix & move the file `spatial_pooler_py_api_test.py` from `nupic.py` into `bindings/py/tests/`.  This tests that the bindings work

Reply:
This is more complicated than I thought. The unit tests don't all pass, they need to be updated...

Reply:
> The unit tests don't all pass, they need to be updated... 

I'd say new PR for that, there will be more things to get the py in line.